File: packages/ragleap-rag/src/ragleap/vectorstores/weaviate_backend.py
"""
Weaviate-backed vector storage for ragleap-rag.

**NOT LIVE-VERIFIED** - no Weaviate instance (cloud or local) was
available to test against during development (same honest caveat as
PineconeBackend/mistral/cohere/voyage). Code-complete against the
actual installed weaviate-client==4.22.0 package's real source code
(every method signature and return-type field used below was
introspected directly, not assumed from documentation) - the current
GA Python client (v3 is deprecated). Treat as best-effort until
confirmed live.

Design notes (same reasoning as PineconeBackend):
- persist_directory= is REQUIRED - a Weaviate Cloud instance persists
  vectors remotely regardless of local process state, so a persistent
  SQLite sidecar for chunk text is mandatory to avoid orphaned vectors.
- A SQLite sidecar stores full chunk text, the document registry, and
  a document_id/chunk_index -> Weaviate UUID mapping (Weaviate assigns
  its own UUIDs on insert - unlike Pinecone, you can't just supply a
  custom string ID directly as the primary key).
- Vectors are "self-provided" (Configure.Vectors.self_provided()) -
  ragleap-rag always brings its own embeddings; Weaviate's built-in
  vectorizer integrations are intentionally not used here, consistent
  with this library's BYOK philosophy.
- supports_sparse() is False even though Weaviate natively supports
  BM25/hybrid search - implementing that natively is a real, valuable
  future enhancement, not done here to keep initial scope honest and
  verifiable rather than guessing at a larger untested surface area.
"""
import json
import logging
import os
import sqlite3
import threading
import uuid as uuid_module
from typing import Dict, List, Optional

from ragleap.vectorstores.base import VectorBackend

logger = logging.getLogger(__name__)


class WeaviateBackend(VectorBackend):
    def __init__(
        self,
        persist_directory: str,
        cluster_url: Optional[str] = None,
        api_key: Optional[str] = None,
        collection_name: str = "RagleapChunk",
    ):
        try:
            import weaviate  # noqa: F401
        except ImportError as e:
            raise ImportError(
                "WeaviateBackend requires the 'weaviate' extra: pip install ragleap-rag[weaviate]"
            ) from e

        if not persist_directory:
            raise ValueError(
                "WeaviateBackend requires persist_directory= - a remote Weaviate "
                "instance persists vectors regardless of local process state, so a "
                "persistent SQLite sidecar for chunk text is mandatory to avoid "
                "orphaned vectors after a restart."
            )

        self.cluster_url = cluster_url or os.environ.get("WEAVIATE_CLUSTER_URL")
        self.api_key = api_key or os.environ.get("WEAVIATE_API_KEY")
        # collection_name must start with an uppercase letter per Weaviate's
        # naming convention - normalize rather than fail confusingly later.
        self.collection_name = collection_name[0].upper() + collection_name[1:] if collection_name else "RagleapChunk"
        self.persist_directory = persist_directory
        self._client = None
        self._collection = None
        self._dimensions = None
        self._lock = threading.Lock()

        os.makedirs(persist_directory, exist_ok=True)
        self._sqlite_path = os.path.join(persist_directory, "weaviate_meta.sqlite3")
        self._conn = sqlite3.connect(self._sqlite_path, check_same_thread=False)
        self._conn.execute(
            """CREATE TABLE IF NOT EXISTS documents (
                id TEXT PRIMARY KEY, filename TEXT NOT NULL, metadata TEXT NOT NULL, uploaded_at TEXT NOT NULL
            )"""
        )
        self._conn.execute(
            """CREATE TABLE IF NOT EXISTS chunks (
                vector_key TEXT PRIMARY KEY, weaviate_uuid TEXT NOT NULL, document_id TEXT NOT NULL,
                document_name TEXT NOT NULL, chunk_index INTEGER NOT NULL,
                text TEXT NOT NULL, token_count INTEGER, metadata TEXT NOT NULL
            )"""
        )
        self._conn.commit()

    def _vector_key(self, document_id: str, chunk_index: int) -> str:
        return f"{document_id}:{chunk_index}"

    def _deterministic_uuid(self, vector_key: str) -> str:
        """Weaviate requires its own UUID as the object identifier -
        deriving it deterministically from our own vector_key (rather
        than a random uuid4) means re-running the same insert is
        idempotent at the Weaviate layer too, matching how FAISS/
        Pinecone's IDs are derived from document_id+chunk_index."""
        return str(uuid_module.uuid5(uuid_module.NAMESPACE_DNS, vector_key))

    def init_schema(self, dimensions: int) -> None:
        import weaviate
        from weaviate.classes.init import Auth
        from weaviate.classes.config import Configure

        self._dimensions = dimensions

        if self.cluster_url:
            auth = Auth.api_key(self.api_key) if self.api_key else None
            self._client = weaviate.connect_to_weaviate_cloud(cluster_url=self.cluster_url, auth_credentials=auth)
        else:
            self._client = weaviate.connect_to_local()

        if not self._client.collections.exists(self.collection_name):
            logger.info(f"WeaviateBackend: creating collection '{self.collection_name}' (dimensions={dimensions})")
            self._client.collections.create(
                name=self.collection_name,
                vector_config=Configure.Vectors.self_provided(),
            )
        else:
            logger.info(f"WeaviateBackend: using existing collection '{self.collection_name}'")

        self._collection = self._client.collections.get(self.collection_name)

    def _build_filter(self, metadata_filter: Optional[Dict]):
        if not metadata_filter:
            return None
        from weaviate.classes.query import Filter

        conditions = [Filter.by_property(k).equal(v) for k, v in metadata_filter.items()]
        combined = conditions[0]
        for cond in conditions[1:]:
            combined = combined & cond
        return combined

    def insert_document(self, document_id: str, filename: str, metadata: Dict) -> None:
        import datetime
        with self._lock:
            self._conn.execute(
                "INSERT INTO documents (id, filename, metadata, uploaded_at) VALUES (?, ?, ?, ?)",
                (document_id, filename, json.dumps(metadata or {}), datetime.datetime.utcnow().isoformat()),
            )
            self._conn.commit()

    def insert_chunk(
        self, document_id: str, document_name: str, chunk_index: int,
        text: str, token_count: int, embedding: List[float], metadata: Dict,
    ) -> None:
        vector_key = self._vector_key(document_id, chunk_index)
        weaviate_uuid = self._deterministic_uuid(vector_key)

        with self._lock:
            self._conn.execute(
                "INSERT INTO chunks (vector_key, weaviate_uuid, document_id, document_name, chunk_index, text, token_count, metadata) "
                "VALUES (?, ?, ?, ?, ?, ?, ?, ?)",
                (vector_key, weaviate_uuid, document_id, document_name, chunk_index, text, token_count, json.dumps(metadata or {})),
            )
            self._conn.commit()

            properties = {
                "document_id": document_id, "document_name": document_name,
                "chunk_index": chunk_index, **(metadata or {}),
            }
            self._collection.data.insert(properties=properties, vector=embedding, uuid=weaviate_uuid)

    def search_dense(self, embedding: List[float], top_k: int, metadata_filter: Optional[Dict] = None) -> List[Dict]:
        if not embedding or self._collection is None:
            return []
        if self._dimensions and len(embedding) != self._dimensions:
            logger.warning(f"Query embedding dim={len(embedding)} != expected {self._dimensions}; skipping search")
            return []

        response = self._collection.query.near_vector(
            near_vector=embedding, limit=top_k,
            filters=self._build_filter(metadata_filter),
            return_metadata=["distance"],
        )

        results = []
        for obj in response.objects:
            row = self._conn.execute(
                "SELECT document_id, document_name, chunk_index, text FROM chunks WHERE weaviate_uuid = ?",
                (str(obj.uuid),),
            ).fetchone()
            if row is None:
                logger.warning(f"WeaviateBackend: object '{obj.uuid}' has no matching local text row, skipping")
                continue
            document_id, document_name, chunk_index, text = row
            # Weaviate returns distance (lower = more similar), not a
            # similarity score directly - convert to a similarity-style
            # score for consistency with the other backends, which all
            # return higher-is-better values.
            distance = obj.metadata.distance if obj.metadata and obj.metadata.distance is not None else None
            similarity_score = round(1.0 - distance, 4) if distance is not None else None
            results.append({
                "chunk_id": str(obj.uuid), "text": text, "similarity_score": similarity_score,
                "document_id": document_id, "document_name": document_name, "chunk_index": chunk_index,
            })
        return results

    def supports_sparse(self) -> bool:
        return False

    def list_documents(self, limit: int, offset: int) -> List[Dict]:
        rows = self._conn.execute(
            """SELECT d.id, d.filename, d.uploaded_at, d.metadata, COUNT(c.vector_key) AS chunk_count
               FROM documents d LEFT JOIN chunks c ON c.document_id = d.id
               GROUP BY d.id ORDER BY d.uploaded_at DESC LIMIT ? OFFSET ?""",
            (limit, offset),
        ).fetchall()
        return [
            {"document_id": r[0], "filename": r[1], "uploaded_at": r[2], "metadata": json.loads(r[3]), "chunk_count": r[4]}
            for r in rows
        ]

    def delete_document(self, document_id: str) -> bool:
        with self._lock:
            weaviate_uuids = [r[0] for r in self._conn.execute(
                "SELECT weaviate_uuid FROM chunks WHERE document_id = ?", (document_id,)
            ).fetchall()]
            if weaviate_uuids and self._collection is not None:
                for wid in weaviate_uuids:
                    self._collection.data.delete_by_id(wid)
            cur = self._conn.execute("DELETE FROM documents WHERE id = ?", (document_id,))
            self._conn.execute("DELETE FROM chunks WHERE document_id = ?", (document_id,))
            deleted = cur.rowcount > 0
            self._conn.commit()
        return deleted

    def get_document_filename(self, document_id: str) -> Optional[str]:
        row = self._conn.execute("SELECT filename FROM documents WHERE id = ?", (document_id,)).fetchone()
        return row[0] if row else None

    def close(self) -> None:
        """Weaviate's v4 client holds an open connection that should be
        explicitly closed when done - unlike Pinecone's REST-based
        client, this isn't automatic. Not called automatically by
        RagLeap; callers managing WeaviateBackend's lifecycle directly
        should call this when finished."""
        if self._client is not None:
            self._client.close()

File: java/ragleap-rag/src/main/java/com/ragleap/rag/vectorstore/WeaviateBackend.java
package com.ragleap.rag.vectorstore;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ArrayNode;
import com.fasterxml.jackson.databind.node.ObjectNode;

import java.io.File;
import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import java.sql.Types;
import java.time.Duration;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.locks.ReentrantLock;

/**
 * Weaviate-backed vector storage for ragleap-rag (Java port).
 *
 * <p>Unlike the Python {@code WeaviateBackend} - which its own source
 * documents as "NOT LIVE-VERIFIED, no Weaviate instance was available to
 * test against during development" - this class and its test suite were
 * built directly against a real, running Weaviate instance from the start.
 * Every endpoint shape, status code, and the exact cosine-distance-to-
 * similarity conversion below were confirmed via live probes before any
 * code was written, not assumed from documentation.
 *
 * <p>Talks to Weaviate's plain REST/GraphQL API via
 * {@link java.net.http.HttpClient} - not the official Java client (if one
 * even fits this project's dependency philosophy), same reasoning as
 * {@link QdrantBackend}. No new Maven dependency needed.
 *
 * <p>Design notes, matching the real Python source's own reasoning where
 * it applies:
 * <ul>
 *   <li><b>persistDirectory and a URL are both required.</b> A remote
 *   Weaviate instance persists vectors regardless of local process state,
 *   so a persistent SQLite sidecar for chunk text is mandatory.</li>
 *   <li><b>Deterministic object IDs</b> via {@link UUID#nameUUIDFromBytes},
 *   derived from {@code documentId:chunkIndex} - Weaviate requires its own
 *   UUID as the object identifier, unlike Pinecone's free-form string IDs.
 *   Deterministic derivation means a document deleted and re-ingested gets
 *   the same Weaviate object ID back.</li>
 *   <li><b>Self-provided vectors</b> ({@code "vectorizer": "none"}) -
 *   ragleap-rag always brings its own embeddings; Weaviate's built-in
 *   vectorizer integrations are intentionally not used, matching this
 *   library's BYOK philosophy.</li>
 *   <li><b>Cosine distance to similarity.</b> Weaviate's GraphQL search
 *   returns {@code distance} (lower is more similar), not a similarity
 *   score. Live-verified: an identical vector returns distance 0, an
 *   orthogonal one returns distance 1 - confirming
 *   {@code similarity = 1 - distance} is correct for the class's cosine
 *   distance metric, which is set explicitly at schema-creation time
 *   here rather than relying on Weaviate's (also-confirmed) default.</li>
 *   <li>{@code supportsSparse()} is {@code false} - Weaviate natively
 *   supports BM25/hybrid search, but implementing that is real future
 *   scope, not guessed at here.</li>
 * </ul>
 */
public class WeaviateBackend implements VectorBackend, AutoCloseable {

    private static final ObjectMapper MAPPER = new ObjectMapper();

    private final String url;
    private final String apiKey;
    private final String collectionName;
    private final Connection conn;
    private final HttpClient http;
    private final ReentrantLock lock = new ReentrantLock();
    private Integer dimensions;

    public WeaviateBackend(String persistDirectory, String url, String apiKey, String collectionName) throws SQLException {
        if (persistDirectory == null || persistDirectory.isBlank()) {
            throw new IllegalArgumentException(
                    "WeaviateBackend requires persistDirectory - a remote Weaviate instance persists "
                            + "vectors regardless of local process state, so a persistent SQLite sidecar "
                            + "for chunk text is mandatory to avoid orphaned vectors after a restart.");
        }

        String resolvedUrl = url != null ? url : System.getenv("WEAVIATE_URL");
        if (resolvedUrl == null || resolvedUrl.isBlank()) {
            throw new IllegalArgumentException(
                    "No url for WeaviateBackend. Pass url explicitly (a Weaviate Cloud cluster URL or a "
                            + "self-hosted instance's address), or set WEAVIATE_URL in your environment.");
        }
        this.url = resolvedUrl;
        this.apiKey = apiKey != null ? apiKey : System.getenv("WEAVIATE_API_KEY");

        // Weaviate class names must start with an uppercase letter -
        // normalize rather than fail confusingly later, matching the
        // Python source's own convention.
        String name = collectionName != null && !collectionName.isBlank() ? collectionName : "RagleapChunk";
        this.collectionName = Character.toUpperCase(name.charAt(0)) + name.substring(1);

        this.http = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(10)).build();

        File dir = new File(persistDirectory);
        if (!dir.exists() && !dir.mkdirs()) {
            throw new IllegalStateException("Could not create persist directory: " + persistDirectory);
        }
        Path sqliteFile = Paths.get(dir.getAbsolutePath(), "weaviate_meta.sqlite3");
        this.conn = DriverManager.getConnection("jdbc:sqlite:" + sqliteFile);
        try (Statement st = conn.createStatement()) {
            st.execute("CREATE TABLE IF NOT EXISTS documents (" +
                    "id TEXT PRIMARY KEY, filename TEXT NOT NULL, metadata TEXT NOT NULL, uploaded_at TEXT NOT NULL)");
            st.execute("CREATE TABLE IF NOT EXISTS chunks (" +
                    "vector_key TEXT PRIMARY KEY, weaviate_uuid TEXT NOT NULL, document_id TEXT NOT NULL, " +
                    "document_name TEXT NOT NULL, chunk_index INTEGER NOT NULL, " +
                    "text TEXT NOT NULL, token_count INTEGER, metadata TEXT NOT NULL)");
        }
    }

    public WeaviateBackend(String persistDirectory, String url) throws SQLException {
        this(persistDirectory, url, null, null);
    }

    /**
     * Package-private, for test verification of the uppercase-first-letter
     * normalization applied in the constructor - not part of the public
     * VectorBackend contract.
     */
    String getCollectionName() {
        return collectionName;
    }

    private String vectorKey(String documentId, int chunkIndex) {
        return documentId + ":" + chunkIndex;
    }

    private String deterministicId(String vectorKey) {
        return UUID.nameUUIDFromBytes(vectorKey.getBytes(StandardCharsets.UTF_8)).toString();
    }

    @Override
    public void initSchema(int dimensions) throws SQLException {
        this.dimensions = dimensions;

        JsonNode existing = request("GET", "/v1/schema/" + collectionName, null, true);
        if (existing == null) {
            ObjectNode vectorIndexConfig = MAPPER.createObjectNode();
            vectorIndexConfig.put("distance", "cosine");

            ObjectNode body = MAPPER.createObjectNode();
            body.put("class", collectionName);
            body.put("vectorizer", "none");
            body.set("vectorIndexConfig", vectorIndexConfig);
            request("POST", "/v1/schema", body, false);
        }
    }

    @Override
    public void insertDocument(String documentId, String filename, Map<String, Object> metadata) throws SQLException {
        String metaJson = toJson(metadata);
        String uploadedAt = OffsetDateTime.now(ZoneOffset.UTC).toString();
        try (PreparedStatement ps = conn.prepareStatement(
                "INSERT INTO documents (id, filename, metadata, uploaded_at) VALUES (?, ?, ?, ?)")) {
            ps.setString(1, documentId);
            ps.setString(2, filename);
            ps.setString(3, metaJson);
            ps.setString(4, uploadedAt);
            ps.executeUpdate();
        }
    }

    @Override
    public void insertChunk(String documentId, String documentName, int chunkIndex, String text,
                             Integer tokenCount, List<Double> embedding, Map<String, Object> metadata) throws SQLException {
        String vectorKey = vectorKey(documentId, chunkIndex);
        String weaviateUuid = deterministicId(vectorKey);
        String metaJson = toJson(metadata);

        lock.lock();
        try (PreparedStatement ps = conn.prepareStatement(
                "INSERT INTO chunks (vector_key, weaviate_uuid, document_id, document_name, chunk_index, text, token_count, metadata) " +
                        "VALUES (?, ?, ?, ?, ?, ?, ?, ?)")) {
            ps.setString(1, vectorKey);
            ps.setString(2, weaviateUuid);
            ps.setString(3, documentId);
            ps.setString(4, documentName);
            ps.setInt(5, chunkIndex);
            ps.setString(6, text);
            if (tokenCount != null) {
                ps.setInt(7, tokenCount);
            } else {
                ps.setNull(7, Types.INTEGER);
            }
            ps.setString(8, metaJson);
            ps.executeUpdate();
        } finally {
            lock.unlock();
        }

        ObjectNode properties = MAPPER.createObjectNode();
        properties.put("document_id", documentId);
        properties.put("document_name", documentName);
        properties.put("chunk_index", chunkIndex);
        if (metadata != null) {
            for (Map.Entry<String, Object> e : metadata.entrySet()) {
                properties.putPOJO(e.getKey(), e.getValue());
            }
        }

        ArrayNode vector = MAPPER.createArrayNode();
        for (Double d : embedding) vector.add(d);

        ObjectNode body = MAPPER.createObjectNode();
        body.put("class", collectionName);
        body.put("id", weaviateUuid);
        body.set("properties", properties);
        body.set("vector", vector);

        request("POST", "/v1/objects", body, false);
    }

    @Override
    public List<SearchResult> searchDense(List<Double> embedding, int topK, Map<String, Object> metadataFilter) throws SQLException {
        if (embedding == null || embedding.isEmpty()) {
            return List.of();
        }
        if (dimensions != null && embedding.size() != dimensions) {
            return List.of();
        }

        StringBuilder vectorLiteral = new StringBuilder("[");
        for (int i = 0; i < embedding.size(); i++) {
            if (i > 0) vectorLiteral.append(",");
            vectorLiteral.append(embedding.get(i));
        }
        vectorLiteral.append("]");

        String whereClause = buildWhereClause(metadataFilter);
        String query = String.format(
                "{ Get { %s(nearVector: {vector: %s}, limit: %d%s) { document_id document_name chunk_index _additional { id distance } } } }",
                collectionName, vectorLiteral, topK, whereClause
        );

        ObjectNode body = MAPPER.createObjectNode();
        body.put("query", query);

        JsonNode response = request("POST", "/v1/graphql", body, false);
        if (response.has("errors") && !response.get("errors").isEmpty()) {
            throw new SQLException("Weaviate GraphQL query returned errors: " + response.get("errors"));
        }

        JsonNode objects = response.path("data").path("Get").path(collectionName);

        List<SearchResult> results = new ArrayList<>();
        for (JsonNode obj : objects) {
            String weaviateUuid = obj.path("_additional").path("id").asText();
            double distance = obj.path("_additional").path("distance").asDouble();

            String documentId, documentName, text;
            int chunkIndex;
            try (PreparedStatement ps = conn.prepareStatement(
                    "SELECT document_id, document_name, chunk_index, text FROM chunks WHERE weaviate_uuid = ?")) {
                ps.setString(1, weaviateUuid);
                try (ResultSet rs = ps.executeQuery()) {
                    if (!rs.next()) continue; // orphaned object, no matching local text row
                    documentId = rs.getString(1);
                    documentName = rs.getString(2);
                    chunkIndex = rs.getInt(3);
                    text = rs.getString(4);
                }
            }

            // Live-verified: distance 0 for an identical vector, distance 1
            // for an orthogonal one - similarity = 1 - distance is correct
            // for this class's cosine distance metric.
            double similarityScore = Math.round((1.0 - distance) * 10000.0) / 10000.0;
            results.add(new SearchResult(weaviateUuid, text, similarityScore, documentId, documentName, chunkIndex));
        }
        return results;
    }

    private String buildWhereClause(Map<String, Object> metadataFilter) {
        if (metadataFilter == null || metadataFilter.isEmpty()) return "";

        List<String> operands = new ArrayList<>();
        for (Map.Entry<String, Object> e : metadataFilter.entrySet()) {
            operands.add(String.format(
                    "{path: [\"%s\"], operator: Equal, %s}",
                    escapeGraphQlString(e.getKey()), graphQlValueField(e.getValue())
            ));
        }

        String where;
        if (operands.size() == 1) {
            where = operands.get(0);
        } else {
            where = "{operator: And, operands: [" + String.join(", ", operands) + "]}";
        }
        return ", where: " + where;
    }

    private String graphQlValueField(Object value) {
        if (value instanceof Boolean b) {
            return "valueBoolean: " + b;
        } else if (value instanceof Integer || value instanceof Long) {
            return "valueInt: " + value;
        } else if (value instanceof Double || value instanceof Float) {
            return "valueNumber: " + value;
        } else {
            return "valueText: \"" + escapeGraphQlString(String.valueOf(value)) + "\"";
        }
    }

    private String escapeGraphQlString(String s) {
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }

    @Override
    public boolean supportsSparse() {
        return false;
    }

    @Override
    public List<DocumentSummary> listDocuments(int limit, int offset) throws SQLException {
        List<DocumentSummary> out = new ArrayList<>();
        String sql = "SELECT d.id, d.filename, d.uploaded_at, d.metadata, COUNT(c.vector_key) AS chunk_count " +
                "FROM documents d LEFT JOIN chunks c ON c.document_id = d.id " +
                "GROUP BY d.id ORDER BY d.uploaded_at DESC LIMIT ? OFFSET ?";
        try (PreparedStatement ps = conn.prepareStatement(sql)) {
            ps.setInt(1, limit);
            ps.setInt(2, offset);
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    out.add(new DocumentSummary(
                            rs.getString(1),
                            rs.getString(2),
                            OffsetDateTime.parse(rs.getString(3)),
                            fromJson(rs.getString(4)),
                            rs.getLong(5)
                    ));
                }
            }
        }
        return out;
    }

    @Override
    public boolean deleteDocument(String documentId) throws SQLException {
        lock.lock();
        try {
            List<String> weaviateUuids = new ArrayList<>();
            try (PreparedStatement ps = conn.prepareStatement("SELECT weaviate_uuid FROM chunks WHERE document_id = ?")) {
                ps.setString(1, documentId);
                try (ResultSet rs = ps.executeQuery()) {
                    while (rs.next()) weaviateUuids.add(rs.getString(1));
                }
            }

            for (String uuid : weaviateUuids) {
                request("DELETE", "/v1/objects/" + collectionName + "/" + uuid, null, true);
            }

            int deletedCount;
            try (PreparedStatement ps = conn.prepareStatement("DELETE FROM documents WHERE id = ?")) {
                ps.setString(1, documentId);
                deletedCount = ps.executeUpdate();
            }
            try (PreparedStatement ps = conn.prepareStatement("DELETE FROM chunks WHERE document_id = ?")) {
                ps.setString(1, documentId);
                ps.executeUpdate();
            }
            return deletedCount > 0;
        } finally {
            lock.unlock();
        }
    }

    @Override
    public Optional<String> getDocumentFilename(String documentId) throws SQLException {
        try (PreparedStatement ps = conn.prepareStatement("SELECT filename FROM documents WHERE id = ?")) {
            ps.setString(1, documentId);
            try (ResultSet rs = ps.executeQuery()) {
                return rs.next() ? Optional.of(rs.getString(1)) : Optional.empty();
            }
        }
    }

    @Override
    public void close() throws SQLException {
        conn.close();
    }

    // --- internals ---

    /**
     * Issues a request against Weaviate's REST/GraphQL API. When
     * allowNotFound is true, a 404 (schema-existence check) or empty 204
     * (object delete) response returns null instead of throwing.
     */
    private JsonNode request(String method, String path, JsonNode body, boolean allowNotFound) throws SQLException {
        try {
            HttpRequest.Builder builder = HttpRequest.newBuilder()
                    .uri(URI.create(url + path))
                    .timeout(Duration.ofSeconds(30))
                    .header("Content-Type", "application/json");
            if (apiKey != null && !apiKey.isBlank()) {
                builder.header("Authorization", "Bearer " + apiKey);
            }
            HttpRequest.BodyPublisher publisher = body != null
                    ? HttpRequest.BodyPublishers.ofString(MAPPER.writeValueAsString(body))
                    : HttpRequest.BodyPublishers.noBody();
            builder.method(method, publisher);

            HttpResponse<String> response = http.send(builder.build(), HttpResponse.BodyHandlers.ofString());

            if (allowNotFound && (response.statusCode() == 404 || response.statusCode() == 204)) {
                return null;
            }
            if (response.statusCode() >= 400) {
                throw new SQLException("Weaviate request failed: " + method + " " + path +
                        " -> HTTP " + response.statusCode() + ": " + response.body());
            }
            if (response.body() == null || response.body().isBlank()) {
                return null;
            }
            return MAPPER.readTree(response.body());
        } catch (IOException | InterruptedException e) {
            if (e instanceof InterruptedException) Thread.currentThread().interrupt();
            throw new SQLException("Weaviate request failed: " + method + " " + path, e);
        }
    }

    private String toJson(Map<String, Object> map) {
        try {
            return MAPPER.writeValueAsString(map != null ? map : Map.of());
        } catch (Exception e) {
            throw new RuntimeException("Failed to serialize metadata", e);
        }
    }

    @SuppressWarnings("unchecked")
    private Map<String, Object> fromJson(String json) {
        try {
            return MAPPER.readValue(json, Map.class);
        } catch (Exception e) {
            throw new RuntimeException("Failed to parse metadata", e);
        }
    }
}
